internal/backend/remote-state/gcs/backend_test.go

```diff
@@ -205,6 +256,119 @@ func setupBackend(t *testing.T, bucket, prefix, key string) backend.Backend {
+func setupKmsKey(t *testing.T, keyDetails map[string]string) string {
+	t.Helper()
+
+	projectID := os.Getenv("GOOGLE_PROJECT")
+	if projectID == "" || os.Getenv("TF_ACC") == "" {
+		t.Skip("This test creates a KMS key ring and key in Cloud KMS. " +
+			"Since this may incur costs, it will only run if " +
+			"the TF_ACC and GOOGLE_PROJECT environment variables are set.")
+	}
+
+	// KMS Client
+	ctx := context.TODO()
+	opts, err := testGetClientOptions(t)
+	if err != nil {
+		e := fmt.Errorf("testGetClientOptions() failed: %s", err)
+		t.Fatal(e)
+	}
+	c, err := kms.NewKeyManagementClient(ctx, opts...)
+	if err != nil {
+		e := fmt.Errorf("kms.NewKeyManagementClient() failed: %v", err)
+		t.Fatal(e)
+	}
+	defer c.Close()
+
+	// Get KMS key ring, create if doesn't exist
+	reqGetKeyRing := &kmspb.GetKeyRingRequest{
+		Name: fmt.Sprintf("projects/%s/locations/%s/keyRings/%s", keyDetails["project"], keyDetails["location"], keyDetails["ringName"]),
+	}
+	var keyRing *kmspb.KeyRing
+	keyRing, err = c.GetKeyRing(ctx, reqGetKeyRing)
+	if err != nil {
+		if !strings.Contains(err.Error(), "NotFound") {
+			// Handle unexpected error that isn't related to the key ring not being made yet
+			t.Fatal(err)
+		}
+		// Create key ring that doesn't exist
+		t.Logf("KMS key ring `%s` not found: creating key ring",
+			fmt.Sprintf("projects/%s/locations/%s/keyRings/%s", keyDetails["project"], keyDetails["location"], keyDetails["ringName"]),
+		)
+		reqCreateKeyRing := &kmspb.CreateKeyRingRequest{
+			Parent:    fmt.Sprintf("projects/%s/locations/%s", keyDetails["project"], keyDetails["location"]),
+			KeyRingId: keyDetails["ringName"],
+		}
+		keyRing, err = c.CreateKeyRing(ctx, reqCreateKeyRing)
+		if err != nil {
+			t.Fatal(err)
+		}
+		t.Logf("KMS key ring `%s` created successfully", keyRing.Name)
+	}
+
+	// Get KMS key, create if doesn't exist (and give GCS service account permission to use)
+	reqGetKey := &kmspb.GetCryptoKeyRequest{
+		Name: fmt.Sprintf("%s/cryptoKeys/%s", keyRing.Name, keyDetails["keyName"]),
+	}
+	var key *kmspb.CryptoKey
+	key, err = c.GetCryptoKey(ctx, reqGetKey)
+	if err != nil {
+		if !strings.Contains(err.Error(), "NotFound") {
+			// Handle unexpected error that isn't related to the key not being made yet
+			t.Fatal(err)
+		}
+		// Create key that doesn't exist
+		t.Logf("KMS key `%s` not found: creating key",
+			fmt.Sprintf("%s/cryptoKeys/%s", keyRing.Name, keyDetails["keyName"]),
+		)
+		reqCreateKey := &kmspb.CreateCryptoKeyRequest{
+			Parent:      keyRing.Name,
+			CryptoKeyId: keyDetails["keyName"],
+			CryptoKey: &kmspb.CryptoKey{
+				Purpose: kmspb.CryptoKey_ENCRYPT_DECRYPT,
+			},
+		}
+		key, err = c.CreateCryptoKey(ctx, reqCreateKey)
+		if err != nil {
+			t.Fatal(err)
+		}
+		t.Logf("KMS key `%s` created successfully", key.Name)
+	}
+
+	// Get GCS Service account email, check has necessary permission on key
+	storageCtx := context.TODO()
+	sc, err := storage.NewClient(storageCtx, opts...) //reuse opts from KMS client
```

Comment:
Note: cannot reuse the backend's storage client (like in the `setupBucket` function) because the KMS key needs to exist before the backend buckets are made.

Reply:
This is a useful note and might be worth inlining as a code comment, too.